arduino/main/Arduino to RPi (PUSH).ino

```diff
@@ -0,0 +1,398 @@
+  //
+
+  int checkNum = buf[len - 3] << 8 | buf[len - 2];
+  //Serial.print("Calc checksum is "); Serial.print(crc16(&buf[1], len - 4)); Serial.print(" vs input checksum is "); Serial.println(checkNum);
+  return (buf[1] & 0b1) && isFrameCorrect && crc16(&buf[1], len - 4) == checkNum && buf[len - 1] == STOP;
+}
+
+// Handshake between Arduino and RPi
+void establishContact() {
+  bool handshake = false;
+  bool expectStopByte = false;
+  // msg[0] = START, msg[1] = numReceive, msg[2] = frame, msg[3] = checkNum, msg[4] = checkNum, msg[5] = STOP
+  byte msg[6];
+  byte i = -1;           // to fill buf
+  while (!handshake) {
+    if (Serial.available() > 0) {
+      msg[++i] = Serial.read();
+
+      if (i == 0 && msg[i] != START) { // If receiving first byte but is not START byte
+        Serial.println("Error, frame doesnt start with 0x7e");
+        i = -1;
+        expectStopByte = false;
+        memset(msg, NULL, 1);
+      }
+      else if (msg[i] == START && !expectStopByte) {  // If is receiving the START byte
+        expectStopByte = true;                        // this is the START byte, the next such byte should be STOP byte
+      }
+      else if (msg[i] == STOP && expectStopByte) {  // If receiving STOP byte
+        Serial.println("Frame terminated, expect receive H-Frame");
+        if (!isFrameCorrect(msg, 'H')) {
+          Serial.println("H-frame invalid, not doing anything");
+          expectStopByte = false;
+          memset(msg, NULL, i + 1);
+          i = -1;
+          continue;
+        }
+        else {
+          // If handshake, repeat message back to primary
+          Serial.print("Returning bytes: ");
+          Serial.write(msg, i + 1);
+          Serial.println("");
+          //Serial.write(msg, i + 1);
+          Serial.println("Success");
+          handshake = true;
+          numReceive = 1;
+
+          memset(msg, NULL, i + 1);
+          i = -1;
+          expectStopByte = false;
+        }
+      }
+      else {  // receiving the other bytes in between
+        Serial.print("byte is ");
+        Serial.write(msg[i]);
+        Serial.println("");
+      }
+      delay(10);
+    }
+  }
+}
+
+void sendIFrame(byte i) {
+  Serial.write(START);
+  byte len = strlen(IFramesBuffer[i]);
+  Serial.write(IFramesBuffer[i][0]);
+  Serial.write(IFramesBuffer[i][1] & 0xFE); // Changes the bit[0] from 1 back to 0
+  for (byte j = 2; j < len; j++) {
+    Serial.write(IFramesBuffer[i][j]);
+  }
+  Serial.write(STOP);
+  Serial.print("Frame number "); Serial.print(i); Serial.print(" = "); Serial.write(IFramesBuffer[i]);
+}
+
+// Reads in the gyro & accel values, packets it into an I-Frame and pushes it to CircularBuffer.
+void ReadValues(void *pvParameters) {
+  unsigned long currentTime;
+  double AcX[3], AcY[3], AcZ[3], GyX[3], GyY[3], GyZ[3];
+  char AcXChar[5], AcYChar[5], AcZChar[5], GyXChar[5], GyYChar[5], GyZChar[5],
+       voltChar[6], currentChar[6], powerChar[6], energyChar[6];
```

Comment:
Try to avoid allocating memory for acxchar, acychar, etc, and put the values from acx etc. directly into the final string. Use sprintf or similar

Reply:
I read that sprintf isn't quite reliable in Arduino, but I'll see what can i do